spring-boot-project/spring-boot/src/main/java/org/springframework/boot/web/server/context/MissingWebServerFactoryBeanFailureAnalyzer.java

```diff
@@ -0,0 +1,54 @@
+/*
+ * Copyright 2012-2022 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.boot.web.server.context;
+
+import org.springframework.boot.WebApplicationType;
+import org.springframework.boot.diagnostics.AbstractFailureAnalyzer;
+import org.springframework.boot.diagnostics.FailureAnalysis;
+import org.springframework.boot.diagnostics.FailureAnalyzer;
+import org.springframework.context.ApplicationContextException;
+
+/**
+ * A {@link FailureAnalyzer} that performs analysis of failures caused by an
+ * {@link MissingWebServerFactoryBeanException}.
+ *
+ * @author Guirong Hu
+ */
+class MissingWebServerFactoryBeanFailureAnalyzer extends AbstractFailureAnalyzer<ApplicationContextException> {
+
+	@Override
+	protected FailureAnalysis analyze(Throwable rootFailure, ApplicationContextException cause) {
+		Throwable rootCause = cause.getCause();
+		if (rootCause instanceof MissingWebServerFactoryBeanException) {
+			WebApplicationType webApplicationType = ((MissingWebServerFactoryBeanException) rootCause)
+					.getWebApplicationType();
+			return new FailureAnalysis(String.format(
+					"Reason: The running web application is of type %s, but the dependent class is missing.",
+					webApplicationType.name().toLowerCase()), action(webApplicationType), cause);
+		}
+		return null;
+	}
+
+	private String action(WebApplicationType webApplicationType) {
+		String action = "Add `%s` dependencies to your application.";
+		if (webApplicationType == WebApplicationType.SERVLET) {
+			return String.format(action, "spring-boot-starter-web");
+		}
+		return String.format(action, "spring-boot-starter-webflux");
```

Comment:
I think this advice is still too specific. For example, a dependency on `spring-boot-starter-webflux` could be in place but the user's set `spring.main.web-application-type` to `servlet`. Another example could be that they have a dependency on `spring-boot-starter-web` with an exclude on `spring-boot-starter-tomcat` but they've forgotten to add a dependency on `spring-boot-starter-jetty` or `spring-boot-starter-undertow`. As I said previously, I don't think we can do much more than suggest that a user:

- Checks their dependencies for a supported web server
- Checks their configuration of web application type. This can be done via a property or programatically so we shouldn't be too prescriptive here either.

Reply:
Thank very much for your guidance @wilkinsona. I've updated the PR.